src/qmapshack/gis/rte/router/brouter/CRouterBRouterDownloadPage.cpp

```diff
@@ -0,0 +1,256 @@
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+**********************************************************************************************/
+
+#include "gis/rte/router/brouter/CRouterBRouterDownloadPage.h"
+#include "gis/rte/router/brouter/CRouterBRouterSetup.h"
+
+#include <JlCompress.h>
+#include <QCheckBox>
+#include <QDir>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QMessageBox>
+#include <QNetworkAccessManager>
+#include <QNetworkReply>
+#include <QPushButton>
+
+CRouterBRouterReleaseItem::CRouterBRouterReleaseItem(const QString &_name, const QString &_description, const QString &_url) : name(_name), description(_description), url(_url) {}
+
+CRouterBRouterDownloadPage::CRouterBRouterDownloadPage() : QWizardPage() {
+  networkAccessManager = new QNetworkAccessManager(this);
+  networkAccessManager->setRedirectPolicy(QNetworkRequest::NoLessSafeRedirectPolicy);
+  connect(networkAccessManager, &QNetworkAccessManager::finished, this,
+          &CRouterBRouterDownloadPage::slotNetworkAccessManagerReplyReceived);
+  releaseItems = new QList<CRouterBRouterReleaseItem>();
+  model = new QStringListModel(this);
+}
+
+CRouterBRouterDownloadPage::~CRouterBRouterDownloadPage() {}
+
+void CRouterBRouterDownloadPage::initialize(QListView* listVersions, QTextBrowser* textVersionDetails, QLabel* labelInstallLink, QPushButton* pushInstall, QTextBrowser* textLocalInstall, CRouterBRouterSetup* setup) {
+  this->listVersions = listVersions;
+  this->textVersionDetails = textVersionDetails;
+  this->labelInstallLink = labelInstallLink;
+  this->pushInstall = pushInstall;
+  this->textLocalInstall = textLocalInstall;
+  this->setup = setup;
+  this->listVersions->setModel(model);
+}
+
+void CRouterBRouterDownloadPage::begin() {
+  textVersionDetails->setVisible(false);
+  textVersionDetails->clear();
+  textLocalInstall->setVisible(false);
+  textLocalInstall->clear();
+  labelInstallLink->setText(tr("no brouter-version to install selected"));
+  pushInstall->setEnabled(false);
+  wizard()->setOption(QWizard::HaveCustomButton1, false);
+  downloadUrl = QUrl();
+
+  QNetworkRequest request(QUrl(setup->expertBinariesUrl));
+  request.setRawHeader(QByteArray("Accept"), QByteArray("application/vnd.github+json"));
+  request.setRawHeader(QByteArray("X-GitHub-Api-Version"), QByteArray("2022-11-28"));
+  QNetworkReply* reply = networkAccessManager->get(request);
+  reply->setProperty("requestType", eRequestVersions);
+}
+
+void CRouterBRouterDownloadPage::slotListVersionsClicked(const QModelIndex& index) {
+  const QString &version = index.data().toString();
+  for(const CRouterBRouterReleaseItem &item : *releaseItems)
```

Comment:
Use `qAsConst` to avoid detaching the container. This applies to all of you ranged loops in this file.

Reply:
fixed